File: test/ast/ast.py
#!/usr/bin/python
#-*- coding: utf-8 -*-
#
# ast test

import os
import sys
import subprocess
import tempfile
import re
from xml.dom import minidom

class TestUnit:
    name = ""
    snippets = []

class TestSnippet:
    code = ""
    expect = ""

def die():
    """fail"""
    exit(1)

def run_snippet(snippet):
    """run snippet"""
    print snippet.code
    print snippet.expect

def run_unit(unit):
    """run unit"""
    print "Testing " + unit.name + " ..."
    for snippet in unit.snippets:
        run_snippet(snippet)

def parse_xml(path):
    """process a xml"""
    testUnit = TestUnit()
    
    dom = minidom.parse(path)
    unit = dom.getElementsByTagName("unit")[0]
    snippets = unit.getElementsByTagName("snippet")
    
    testUnit.name = unit.attributes["name"].value
    
    for snippet in snippets:
        testSnippet = TestSnippet()
        code = snippet.getElementsByTagName("code")[0]
        expect = snippet.getElementsByTagName("expect")[0]
        testSnippet.code = code.firstChild.nodeValue
        testSnippet.expect = expect.firstChild.nodeValue;
        testUnit.snippets.append(testSnippet)

    return testUnit

def main():
    """main"""
    print
    dir = os.path.dirname(sys.argv[0])
    xmls = [f for f in os.listdir(dir) if f.endswith(".xml")]
    for xml in xmls:
        path = os.path.join(dir, xml)
        unit = parse_xml(path)
        run_unit(unit)
    
    return 0
    
if __name__ == "__main__":
    reload(sys)
    sys.setdefaultencoding('utf8')
    sys.exit(main())

File: test/ast/ast.rb
#!/usr/bin/ruby
#
# test

require 'rexml/document'
require 'Pathname'
require 'open3'
require 'tmpdir'

$XML
$UNIT
$CASE

def fail
    puts "\033[1;31mFAILED\033[0m"
    puts "#{$XML}:unit #{$UNIT}:case #{$CASE}"
    exit 1
end

def ok
   puts "\033[32mOK\033[0m"
end

def cc_code(code)
    path = Dir.mktmpdir("mcc") + "/1.c"
    f = File.open(path, "w")
    f.puts code
    f.close
    stdout, stderr, status = Open3.capture3("./mcc " + path)
    if status != 0 then fail end
    stderr
end

def to_xml(output)
    puts output
end

def diff(output, expect)
    
end

def run_case(e)
    code = e.elements.find { |k| k.name == "code" }
    expect = e.elements.find { |k| k.name == "expect" }
    output = cc_code code.text
    diff(to_xml(output), expect)
end

def run_unit(unit)
    $UNIT = unit.attributes["name"]
    if $UNIT == nil
       $UNIT = File.basename($XML, File.extname($XML))
    end
    print "Testing " + $UNIT + " ..."
    cases = unit.elements.select { |k| k.name == "case" }
    cases.each do |e|
        $CASE = unit.index e
        run_case e
    end
end

def parse_xml(path)
    (REXML::Document.new File.open path).root
end

def main
    dir = Pathname.new(__FILE__).dirname
    xmls = Dir.glob dir + "*.xml"
    xmls.each do |xml|
        path = Pathname.pwd + xml
        $XML = xml
        run_unit parse_xml path
        ok
    end
end

main